epp-client/src/domain/check.rs

```diff
@@ -0,0 +1,146 @@
+use epp_client_macros::*;
+
+use serde::{Deserialize, Serialize};
+
+use crate::epp::object::{ElementName, EmptyTag, StringValue};
+use crate::epp::request::EppRequest;
+use crate::epp::xml::EPP_DOMAIN_XMLNS;
+
+/// Type that represents the &lt;epp&gt; request for domain &lt;check&gt; command
+///
+/// ## Usage
+///
+/// ```no_run
+/// use std::collections::HashMap;
+///
+/// use epp_client::config::{EppClientConfig, EppClientConnection};
+/// use epp_client::EppClient;
+/// use epp_client::domain::check::Check;
+/// use epp_client::epp::generate_client_tr_id;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     // Create a config
+///     let mut registry: HashMap<String, EppClientConnection> = HashMap::new();
+///     registry.insert(
+///         "registry_name".to_owned(),
+///         EppClientConnection {
+///             host: "example.com".to_owned(),
+///             port: 700,
+///             username: "username".to_owned(),
+///             password: "password".to_owned(),
+///             ext_uris: None,
+///             tls_files: None,
+///         },
+///     );
+///     let config = EppClientConfig { registry };
+///
+///     // Create an instance of EppClient, passing the config and the registry you want to connect to
+///     let mut client = match EppClient::new(&config, "registry_name").await {
+///         Ok(client) => client,
+///         Err(e) => panic!("Failed to create EppClient: {}",  e)
+///     };
+///
+///     // Create a Check instance
+///     let domain_check = Check::new(vec!["eppdev-100.com", "eppdev-100.net"];
+///
+///     // send it to the registry and receive a response of type DomainCheck
+///     let response = client.transact_new::<_, DomainCheck>(&domain_check).await.unwrap();
+///
+///     println!("{:?}", response);
+///
+///     client.logout().await.unwrap();
+/// }
+/// ```
+
+#[derive(Debug)]
+pub struct DomainCheck {
+    request: DomainCheckRequest,
+}
+
+impl EppRequest for DomainCheck {
```

Comment:
Hi @nrempel ,

Thanks for working on this reorganization. I need some help with understanding how these changes might work when using the library.

From what I understand from this `impl` block for `DomainCheck`

```rust
impl EppRequest for DomainCheck {
    type Input = DomainCheckRequest;
    type InputExtension = EmptyTag;
    type Output = DomainCheckResponse;
    type OutputExtension = EmptyTag;
}
```

To add a new type of extension, like maybe the namestore subproduct extension for domain check, we wouldn't be able to re-use this `DomainCheck` type since the `EppRequest` implementation has the associated extension type already set to `EmptyTag`? Would we have to make a new type in that case for each domain check extension? What would the syntax look like for the `transact()` function since in this doc example, it needs typeargs which we were trying to eliminate from the `transact()` call?

Reply:
Hi, yes in this design we would have needed to re-implement each of the requests separately to include the extension. I much prefer your approach to setting the extension dynamically. I took a shot at a hybrid approach, more details below.